Fix duplicate metrics in prometheus

alim4r opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
Every prometheus metric has at least 3 labels (`entity`, `name`, `domain`).
If a user changes the `friendly_name` a new metric is created, but the old one keeps existing. This was reported in #50382.
This also happens when an entity id is renamed and by deleting or disabling an entity.

In this pull request the `EVENT_ENTITY_REGISTRY_UPDATED` event is used to listen to the changes and to remove the non-existing labelsets.
To remove a labelset the function has to iterate over all labelsets. Since this only happens when an entity is renamed, disabled or deleted this shouldn't make a difference in performance.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #50382
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

alim4r committed:
Fix duplicate metrics in prometheus

alim4r commented:
I am not sure how to fix the code coverage error: `Added lines #L226 - L227 were not covered by tests`
I wasn't sure what would happen if two events wanted to remove a labelset in an async context and added the catch for `KeyError` as a sanity check.
Should I remove the `try catch` or is there another way to test this case?

https://github.com/home-assistant/core/blob/f69fd3dc97fdafb746f286c7e36d3fae7670e3ca/homeassistant/components/prometheus/__init__.py#L214-L228

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -185,6 +189,44 @@ def handle_event(self, event):
+                        metric.remove(*sample.labels.values())
+                    except KeyError:
+                        pass
+                    break
```
I believe this will result in only the first labelset with a given entity label set being removed, while we need to remove all of them.

alim4r replied:
Good catch! I didn't think of that. The `break` could be removed or the functions checks if the metric name contains `climate_action` or `humidifier_mode`.

1. If the `break` is removed the `remove` functions will be called multiple times but the `KeyError` will be catched (for example for Counter Metrics: `state_change` with child metrics `state_change_created` and `state_change_total`)

2. If the metric name is checked the function has to be updated every time a new metric with multiple labelsets for the same entity is added

I think option 1 would be better?

In both cases the `friendly_name` should be added as a parameter to `_remove_labelsets` since it no longer removes just the oldest labelset. `handle_state_changed` could be faster than `handle_entity_registry_updated` and mistakenly remove the newly generated metric. With the combination of `entity_id` and `friendly_name` it should be unique.

alim4r replied:
I updated `_remove_labelsets` with a `friendly_name` parameter and removed the `break` command.
The tests randomly failed from time to time. When it fails `entity_registry_updated` event returns returns `changes=name=None`  instead of `changes=name="Outside Temperature"` . I assume its a timing issue.
Because of that I moved the `friendly_name` update detection to `handle_state_changed` which seems to work as expected, but doesn't look as clean since its now handled in two events.

## Log
````
DEBUG:homeassistant.core:Bus:Handling <Event entity_registry_updated[L]: action=create, entity_id=sensor.outside_temperature>
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=sensor.outside_temperature, old_state=None, new_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Outside Temperature @ 2021-12-09T17:52:05.071434+00:00>>

DEBUG:homeassistant.core:Bus:Handling <Event entity_registry_updated[L]: action=update, entity_id=sensor.outside_temperature, changes=name=None>
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=sensor.outside_temperature, old_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Outside Temperature @ 2021-12-09T17:52:05.071434+00:00>, new_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Inside Temperature @ 2021-12-09T17:52:05.071434+00:00>>
```

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -639,6 +640,351 @@ async def test_lock(hass, hass_client):
+    assert "sensor.outside_temperature" in registry.entities
+    registry.async_update_entity(
+        entity_id="sensor.outside_temperature",
+        name="Inside Temperature",
```
Perhaps `name="Outside Temperature Renamed"` will make it easier to understand what this test is checking?

alim4r replied:
Thanks. I updated it in the new commit.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -639,6 +640,351 @@ async def test_lock(hass, hass_client):
+    body = await generate_latest_metrics(client)
+
+    # Check if old metrics deleted
+    assert (
+        'sensor_temperature_celsius{domain="sensor",'
+        'entity="sensor.outside_temperature",'
+        'friendly_name="Outside Temperature"} 15.6' not in body
+    )
+
+    assert (
+        'entity_available{domain="sensor",'
+        'entity="sensor.outside_temperature",'
+        'friendly_name="Outside Temperature"} 1.0' not in body
+    )
```
I would suggest a broader check here, perhaps just `assert 'friendly_name="Outside Temperature"' not in '\n'.join(body)`?

This applies to all other similar cases below.

alim4r replied:
Thanks, this seems to be the better solution! Added it in the new commit.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -639,6 +640,351 @@ async def test_lock(hass, hass_client):
+    )
+
+    # Keep other sensors
+    assert (
+        'sensor_humidity_percent{domain="sensor",'
+        'entity="sensor.outside_humidity",'
+        'friendly_name="Outside Humidity"} 54.0' in body
+    )
+
+    assert (
+        'entity_available{domain="sensor",'
+        'entity="sensor.outside_humidity",'
+        'friendly_name="Outside Humidity"} 1.0' in body
+    )
+
```
Again, perhaps this can be simplified with a broader check, like `assert 'friendly_name="Outside Humidity"' in '\n'.join(body)`?

alim4r replied:
This would stop at the first match. If one of the metrics would be deleted by mistake, the test wouldn't notice it.
On the other hand we're not checking every metric for `sensor.outside_humidity` at the moment anyway.
Should I change it regardless?

knyar replied:
Oh yes, you are right, it's safer to have more specific checks here.

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -185,6 +189,44 @@ def handle_event(self, event):
+            changes = event.data.get("changes")
+            if "entity_id" in changes:
+                metrics_entity_id = changes["entity_id"]
+            elif "disabled_by" in changes or "name" in changes:
```
For completion, should we also react on updates to `entity_category`?

alim4r replied:
I don't think so. Since `entity_category` is not used in the labelsets, it shouldn't affect the metrics?

knyar replied:
Oh sorry, I meant `domain` rather than `entity_category`. I suspect we don't need to cover it since entities cannot move from one domain to another, right?

knyar reviewed (changes requested):
Thank you for proposing this! Looks good overall, just a few minor comments.

> I wasn't sure what would happen if two events wanted to remove the same labelset in an async context and added the catch for KeyError as a sanity check.
> Should I remove the try catch or is there another way to test this case?

I wonder if we should add a lock protecting changes to `self._metrics` and make `_remove_labelsets` take it before it iterates over `self._metrics`, which will guarantee that we don't try to remove labelsets concurrently.

Having such a lock might also fix a potential race in `_metric()` where two concurrent threads might create two separate Prometheus metrics with only one of them assigned to `self._metrics`

alim4r committed:
Fix duplicate prometheus metrics for entities with multiple labelsets
- Move friendly_name detection to state_changed event
- Add additional test case

alim4r committed:
Add review suggestions for prometheus friendly name update

alim4r committed:
Remove commented out code in prometheus

alim4r committed:
Update prometheus tests for deleted metrics

knyar reviewed (approved):
Thank you, this looks good.

Hopefully one of the Home Assistant maintainers will be able to merge this soon.

alim4r commented:
I rebased the pull request to resolve conflicts with upstream changes

allenporter commented on tests/components/prometheus/test_init.py:
```diff
@@ -712,6 +713,392 @@ async def test_counter(hass, hass_client):
+    )
+    client = await setup_prometheus_client(hass, hass_client, "")
+
+    await async_setup_component(
```
assert here too?

alim4r replied:
Thanks, I added `assert` in the new commit.

allenporter reviewed (commented):
This looks great, just some minor suggestions.

alim4r committed:
Add review suggestions for prometheus
- Remove unnecessary firendly_name check in handle_entity_registry_updated
- Add assert in test

alim4r commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -188,6 +197,48 @@ def handle_event(self, event):
+            elif "disabled_by" in changes:
+                metrics_entity_id = entity_id
+            elif "name" in changes:
+                return
```
You are right. I moved `friendly_name` change detection to  `handle_state_changed` and kept the check. I removed it in the new commit.
